src/chef_index.erl

```diff
@@ -0,0 +1,149 @@
+    end.
+
+-spec query_from_params(binary() | [byte()],
+                        undefined | nonempty_string(),
+                        string(),
+                        string()) -> #chef_solr_query{}.
+query_from_params(ObjType, QueryString, Start, Rows) ->
+    chef_index_query:from_params(search_provider(), ObjType, QueryString, Start, Rows).
+
+-spec add_org_guid_to_query(#chef_solr_query{}, binary()) -> #chef_solr_query{}.
+add_org_guid_to_query(Query, OrgGuid) ->
+    chef_index_query:add_org_guid_to_query(Query, OrgGuid).
+
+transform_data(Data) ->
+    transform_data(search_provider(), Data).
+
+transform_data(elasticsearch, Data) ->
+    chef_elasticsearch:transform_data(Data);
+transform_data(solr, Data) ->
+    chef_solr:transform_data(Data).
+
+%% The difference between add and add_async is that add_async
+%% will not block on the data being posted to our search store
+%% and will instantly return `ok` (regardless of queue mode);
+%% whereas, add will block the request until the queue has been
+%% flushed to our search store in batch queue mode or directly
+%% posted in inline queue mode.
+add(TypeName, Id, DbName, IndexEjson, ReqId) ->
+    QueueMode = queue_mode(),
+    case QueueMode of
+        rabbitmq ->
+            ok = chef_index_queue:set(envy:get(chef_index, rabbitmq_vhost, binary), TypeName, Id, DbName, IndexEjson);
+        _ -> %% else batch or inline, create doc
+            TypeName2 = case TypeName of
+                            data_bag_item ->
+                                ej:get({<<"data_bag">>}, IndexEjson);
+                            T ->
+                                T
+                        end,
+            Doc = chef_index_expand:doc_for_index(TypeName2, Id, DbName, IndexEjson),
+            send_to_solr(QueueMode, Doc, ReqId)
+    end.
+
+add_async(TypeName, Id, DbName, IndexEjson, _ReqId) ->
+    %% No solr_time logging for add_async as this doesn't happen as
+    %% part of a request, But we provided a add_async/5 to make it
+    %% easier to call in oc_chef_object_db.
+    add_async(TypeName, Id, DbName, IndexEjson).
+
+add_async(TypeName, Id, DbName, IndexEjson) ->
+    spawn(chef_index, add, [TypeName, Id, DbName, IndexEjson, none]),
+    ok.
+
+delete(TypeName, Id, DbName, ReqId) ->
+    case queue_mode() of
+        rabbitmq ->
+            ok = chef_index_queue:delete(envy:get(chef_index, rabbitmq_vhost, binary), TypeName, Id, DbName);
+        _ -> %% batch mode not implemented for delete, always use inline if not rabbitmq
+            stats_hero:ctime(ReqId, {chef_solr, delete},
+                             fun() ->
+                                     Doc = chef_index_expand:doc_for_delete(TypeName, Id, DbName),
+                                     chef_index_expand:send_delete(Doc)
+                             end)
+    end.
+
+queue_mode() ->
+    envy:get(chef_index, search_queue_mode, rabbitmq, envy:one_of([rabbitmq, batch, inline])).
```

Comment:
Do we think there would be any benefit of completely excising the batch mode from this code?